airbyte-integrations/connectors/source-genesys/source_genesys/source.py

```diff
@@ -0,0 +1,308 @@
+        return "users"
+
+
+class UserGroups(GenesysStream):
+    """
+    API Docs: https://developer.genesys.cloud/useragentman/groups/
+    """
+
+    primary_key = "id"
+    cursor_field = "dateModified"
+
+    def path(self, **kwargs) -> str:
+        return "groups"
+
+
+class SourceGenesys(AbstractSource):
+    @staticmethod
+    def get_connection_response(self, config: Mapping[str, Any]):
+        GENESYS_TENANT_ENDPOINT_MAP: Dict = {
+            "Americas (US East)": "https://login.mypurecloud.com",
+            "Americas (US East 2)": "https://login.use2.us-gov-pure.cloud",
+            "Americas (US West)": "https://login.usw2.pure.cloud",
+            "Americas (Canada)": "https://login.cac1.pure.cloud",
+            "Americas (São Paulo)": "https://login.sae1.pure.cloud",
+            "EMEA (Frankfurt)": "https://login.mypurecloud.de",
+            "EMEA (Dublin)": "https://login.mypurecloud.ie",
+            "EMEA (London)": "https://login.euw2.pure.cloud",
+            "Asia Pacific (Mumbai)": "https://login.aps1.pure.cloud",
+            "Asia Pacific (Seoul)": "https://login.apne2.pure.cloud",
+            "Asia Pacific (Sydney)": "https://login.mypurecloud.com.au",
+        }
+
+        token_refresh_endpoint = GENESYS_TENANT_ENDPOINT_MAP.get(config["tenant_endpoint"])
+        token_refresh_endpoint = token_refresh_endpoint + "/oauth/token"
+        client_id = config["client_id"]
+        client_secret = config["client_secret"]
+        refresh_token = None
+        headers = {"content-type": "application/x-www-form-urlencoded"}
+        data = {"grant_type": "client_credentials", "client_id": client_id, "client_secret": client_secret, "refresh_token": refresh_token}
+
+        try:
+            response = requests.request(method="POST", url=token_refresh_endpoint, data=data, headers=headers)
+        except Exception as e:
+            raise Exception(f"Error while refreshing access token: {e}") from e
+        return response
+
+    def check_connection(self, logger, config) -> Tuple[bool, any]:
+        """
+        Testing connection availability for the connector by granting the credentials.
+        """
+
+        try:
+            if not config["client_secret"] or not config["client_id"]:
+                raise Exception("Empty config values! Check your configuration file!")
+
+            self.get_connection_response(config).raise_for_status()
+            return True, None
+
+        except Exception as e:
+            return (
+                False,
+                f"Got an exception while trying to set up the connection: {e}. "
+                f"Most probably, there are no users in the given Genesys instance or your token is incorrect",
+            )
+
+    def streams(self, config: Mapping[str, Any]) -> List[Stream]:
+        response = self.get_connection_response(self, config)
+        response.raise_for_status()
+
+        args = {"authenticator": TokenAuthenticator(response.json()["access_token"])}
```

Comment:
The less custom code is, the easier to maintain the airbyte platform. Would it be easier if you derive the class [Oauth2Authenticator](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py)? The derived class only overrides one method -- `build_refresh_request_body`. 

```
def build_refresh_request_body(self) -> Mapping[str, Any]:
          return {
              "grant_type": "client_credentials",
              "client_id": self.get_client_id(),
              "client_secret": self.get_client_secret(),
          }
```

As a result, you don't need to implement get_connection_response. 

Since the tenant depends on the region, you probably still need the mapping and set `token_refresh_endpoint`, in order to instantiate an OAuth2Authenticator. 

```
def __init__(self, domain: str, client_id: str, client_secret: str):
        super().__init__(parse.urljoin(domain, "/oauth/token"), client_id, client_secret, "")
```

So, this is how it's instantiated.  
```
domain = GENESYS_TENANT_ENDPOINT_MAP.get(config["tenant_endpoint"])
args = {"authenticator": GenesysOAuthAuthenticator(domain, config["client_id"], config["client_secret"])}
```

Reply:
@koconder , trying to follow up on the change request, do you have a question? 

I encountered the same problem, and this is how I override the [Oauth2Authenticator](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py) . 

https://github.com/airbytehq/airbyte/pull/18338/files#diff-eed6d6f693f2eb118ec49fdabe99735abb85542039852a4f0e3cc91d8758c188R14

Reply:
@YiyangLi i managed to use your example to extend the original Oauth2Authenticator :)